preprocess.py

```diff
@@ -3,91 +3,220 @@
+    # We've empty'ed each Dataset's `fields` attribute when saving,
+    # so restore them.
+    for train in train_list:
+        train.fields = fields
+
+    onmt.IO.build_vocab(train_list, opt.share_vocab,
+                        opt.src_vocab_size,
+                        opt.src_words_min_frequency,
+                        opt.tgt_vocab_size,
+                        opt.tgt_words_min_frequency)
 
-# **Preprocess Options**
-parser.add_argument('-config', help="Read options from this file")
-
-parser.add_argument('-data_type', default="text",
-                    help="Type of the source input. Options are [text|img].")
-parser.add_argument('-data_img_dir', default=".",
-                    help="Location of source images")
-
-parser.add_argument('-train_src', required=True,
-                    help="Path to the training source data")
-parser.add_argument('-train_tgt', required=True,
-                    help="Path to the training target data")
-parser.add_argument('-valid_src', required=True,
-                    help="Path to the validation source data")
-parser.add_argument('-valid_tgt', required=True,
-                    help="Path to the validation target data")
-
-parser.add_argument('-save_data', required=True,
-                    help="Output file for the prepared data")
-
-parser.add_argument('-src_vocab',
-                    help="Path to an existing source vocabulary")
-parser.add_argument('-tgt_vocab',
-                    help="Path to an existing target vocabulary")
-parser.add_argument('-features_vocabs_prefix', type=str, default='',
-                    help="Path prefix to existing features vocabularies")
-parser.add_argument('-seed', type=int, default=3435,
-                    help="Random seed")
-parser.add_argument('-report_every', type=int, default=100000,
-                    help="Report status every this many sentences")
+    # Can't save fields, so remove/reconstruct at training time.
+    torch.save(onmt.IO.save_vocab(fields),
+               open(opt.save_data + '.vocab.pt', 'wb'))
 
-opts.preprocess_opts(parser)
 
-opt = parser.parse_args()
-torch.manual_seed(opt.seed)
+def build_save_dataset(data_type, fields, opt):
+    assert data_type in ['train', 'valid']
+
+    if data_type == 'train':
+        src_corpus = opt.train_src
+        tgt_corpus = opt.train_tgt
+    else:
+        src_corpus = opt.valid_src
+        tgt_corpus = opt.valid_tgt
+
+    corpus_size = os.path.getsize(src_corpus)
+    if corpus_size > 10 * (1024**2) and opt.max_shard_size == 0:
+        print("Warning! The corpus %s is larger than 10M bytes, you can "
+              "set '-max_shard_size' to divide it into small shards "
+              "to avoid memory hogging problem !!!" % src_corpus)
+
+    # If max_shard_size == 0 or the max_shard_size is bigger than the corpus,
+    # we don't divide the corpus, just preprocess it monolithically.
+    if opt.max_shard_size == 0 or opt.max_shard_size >= corpus_size:
```

Comment:
We should always use the same code path when possible. If you need to special case `max_shard_size` just special case the file name.